src/cpu/o3/store_set.hh

```diff
@@ -34,12 +34,73 @@
 #include <utility>
 #include <vector>
 
+#include "base/cache/associative_cache.hh"
 #include "base/types.hh"
 #include "cpu/inst_seq.hh"
+#include "mem/cache/tags/indexing_policies/base.hh"
+#include "params/SSITIndexingPolicy.hh"
+#include "params/SSITSetAssociative.hh"
 
 namespace gem5
 {
 
+class SSITTagTypes
```

Comment:
There is no need for this any of these new classes, as they are entirely duplicated code. You should use SetAssociative (src/mem/cache/tags/indexing_policies/set_associative.hh) and Addr instead.

Reply:
I cannot use SetAssociative from mem-cache as that is organized based on cache size whereas store set is mainly organized by the number of entries. And because of this, I cannot use AddrTypes. I need to define my own SSITTagType to use the right indexing policy param.

I tried adding another constructor to SetAssociative that takes 'num_entries' but couldn't make it work.

Please let me know if you think we should still be able to use the existing classes.

Reply:
I see the following in the python file:
```
SSITSetAssociative(
            num_entries=Parent.SSITSize,
```

Can´t we just do:
```
SetAssociative(
            size=Parent.SSITSize,
            entry_size=1,
```
?

Reply:
We need to shift the PC to ignore the lower 2 bits so as to index the SSIT using instruction granularity (esp. for ARM). If we set entry_size to 1, we don't do this.

Reply:
Then we adjust the parameters accordingly, as that is exactly what those params are for:
SetAssociative(
            size=Parent.SSITSize * 4,
            entry_size=4,

Reply:
Here 'size' is of MemoryType. We have to declare SSITSize also as a memory type. If we do this, we cannot multiply 'SSITSize * 4' as they are of different types. Do you know how to get this to work?

Reply:
@odanrc Do you have any further comments on this patch? Please let me know since I want to get this merged for the next imminent release. Thank you!